core/internal/filetransfer/file_transfer_gcs.go

```diff
@@ -54,5 +77,228 @@ func (ft *GCSFileTransfer) Upload(task *ReferenceArtifactUploadTask) error {
 func (ft *GCSFileTransfer) Download(task *ReferenceArtifactDownloadTask) error {
 	ft.logger.Debug("GCSFileTransfer: Download: downloading file", "path", task.Path, "ref", task.Reference)
 
-	return fmt.Errorf("not implemented yet")
+	reference := task.Reference
+
+	// Parse the reference path to get the scheme, bucket, and object
+	bucketName, objectName, err := parseReference(reference)
+	if err != nil {
+		return formatDownloadError("error parsing reference", err)
+	}
+	bucket := ft.client.Bucket(bucketName)
+
+	// If not using checksum, we set the digest to the reference
```

Comment:
what does this mean?

Reply:
we allow uploading a reference with a flag `checksum=False`, which will basically just find everything under that reference path (so if its a directory, all the items in the directory) without storing checksums for the files. then when we download it, we just download everything under that path without verifying checksums/etags. 

in the artifact manifest, we would normally store the file checksum under the `digest` field. when uploading with `checksum=False`, we set `digest` as the reference path, and so we just check if the two are equal to know that this entry was uploaded in this way

Reply:
All this should be documented and/or encoded in the `ReferenceArtifactTask` struct. The current doc for `Digest` is

> Digest is the checksum to ensure the correct files are being downloaded

which is false, as it might not be a checksum, and it's missing a description of the relationship with `Reference` (which is that `Digest == Reference` iff the artifact was uploaded with checksum=False) and the fact that it can be used to infer whether the artifact reference is necessarily a single file.

It would be best to encode these rules as `ReferenceArtifactTask` methods. Instead of `task.Digest == reference`, the code should read

```go
var objectNames []string

if !task.MayHaveMultipleFiles() {
	objectNames = []string{objectPathPrefix}
} else {
	objectNames = ft.listObjectNamesWithPrefix(objectPathPrefix)
}

ft.downloadFiles(objectNames)
```

And the checksumming portion should look like

```go
if task.ShouldCheckDigest() && objAttrs.Etag != task.Digest {
	// return error
}
```